### PR TITLE
Reduce crank and crossplane binary size

### LINKED ISSUE #5986: Reduce Binary size (remove debug information)
### What problem are you facing?

The current binaries are being distributed with all the debug symbols. They can be removed to significantly reduce the binary size

Current size for v1.17.1 on MacOS ARM:
1. Crank: 80M
2. Crossplane: 73M

### PR DESCRIPTION
**Pull request by Moulick**
### Description of your changes

Remove debug symbols from binaries

Fixes #5986

This reduces the size:
1. Crank: 80M -> 55M (~31% reduction)
2. Crossplane: 73M -> 51M (~30% reduction)

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing

### CODE REVIEW

File: Earthfile [hunk 1]
```diff
@@ -151,7 +151,7 @@ go-build:
   ARG TARGETOS
   ARG GOARCH=${TARGETARCH}
   ARG GOOS=${TARGETOS}
-  ARG GOFLAGS="-ldflags=-X=github.com/crossplane/crossplane/internal/version.version=${CROSSPLANE_VERSION}"
+  ARG GOFLAGS="-ldflags=-s -w -X=github.com/crossplane/crossplane/internal/version.version=${CROSSPLANE_VERSION}"
```
Comment (Moulick) [inline Earthfile:154]: Is this the only place where `go build` happens for release binaries?

**Comment by negz** [comment]
@Moulick What tradeoff are we making here - is there anything we lose by dropping the debug symbols?

**Comment by Moulick** [comment]
> @Moulick What tradeoff are we making here - is there anything we lose by dropping the debug symbols?

None as far as I know. The DWARF and go debug table are mainly useful for debuggers. While reading a little more about this topic I found another flag `-trimpath`, which when I tried locally shaved off another MB. But `-trimpath` is only necessary if the build is being done locally to remove compromising information such as the name of the compiling user. IMHO not necessary for binaries being built in CI.

https://xnacly.me/posts/2023/go-metadata/

**Review by jbw976 — COMMENTED** [review]
@Moulick thanks for the initiative you continue to take in the community, really cool to see you keep contributing! 🙇‍♂️ 

do you happen to know what happens in the following scenarios?

* a panic occurs - does the stack trace still get included in the pod's output or terminal output?
* some part of controller-runtime wants to print the stack, like we've seen in multiple places e.g. https://github.com/crossplane-contrib/provider-upjet-aws/issues/973 - is it still possible to see that stack output?

It may be interesting to purposefully inject a panic and test this out so we see concretely 🤔

**Comment by Moulick** [comment]
So added a panic to the version function as below and here are the stacktrace for the panic with various flags

```diff
// Run runs the version command.
func (c *Cmd) Run(k *kong.Context) error {
	_, _ = fmt.Fprintln(k.Stdout, "Client Version: "+version.New().GetVersionString())
	if c.Client {
		return nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	vxp, err := FetchCrossplaneVersion(ctx)
+	panic("Trim the fat test")
	if err != nil {
		return errors.Wrap(err, errGetCrossplaneVersion)
	}
	if vxp != "" {
		_, _ = fmt.Fprintln(k.Stdout, "Server Version: "+vxp)
	}

	return nil
}
```

Here are the stacktrace for 

1. As it is now

```shell
$ go build -o crank -ldflags='-X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version
```

<details>
<summary>Panic</summary>

```shell
$ go build -o crank -ldflags='-X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version

Client Version: v1.2.3
panic: Trim the fat test

goroutine 1 [running]:
github.com/crossplane/crossplane/cmd/crank/version.(*Cmd).Run(0x1400052e4b8, 0x140003f1b00)
	/Users/moulick.aggarwal/redacted/crossplane/cmd/crank/version/version.go:51 +0x104
reflect.Value.call({0x104c6e320?, 0x1400052e4b8?, 0x14000069aa8?}, {0x1046abd1e, 0x4}, {0x14000780e70, 0x1, 0x102f08f88?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/reflect/value.go:581 +0x97c
reflect.Value.Call({0x104c6e320?, 0x1400052e4b8?, 0x14000069ba8?}, {0x14000780e70?, 0x10500d5c0?, 0x102df9d58?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/reflect/value.go:365 +0x94
github.com/alecthomas/kong.callFunction({0x104c6e320?, 0x1400052e4b8?, 0x14000069d08?}, 0x14000702c00)
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/callbacks.go:98 +0x390
github.com/alecthomas/kong.(*Context).RunNode(0x140003f1b00, 0x140005d10e0, {0x0, 0x0, 0x140005b8780?})
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/context.go:774 +0x604
github.com/alecthomas/kong.(*Context).Run(0x104b38560?, {0x0?, 0x14000069ee8?, 0x10000000000013?})
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/context.go:799 +0x138
main.main()
	/Users/moulick.aggarwal/redacted/crossplane/cmd/crank/main.go:77 +0x240
```

</details>

2. With `ldflags -s -w`

```shell
$ go build -o crank -ldflags='-s -w -X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version
```

<details>
<summary>Panic</summary>

Client Version: v1.2.3
panic: Trim the fat test

```shell
$ go build -o crank -ldflags='-s -w -X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version

goroutine 1 [running]:
github.com/crossplane/crossplane/cmd/crank/version.(*Cmd).Run(0x1400032f3b8, 0x14000817300)
	/Users/moulick.aggarwal/redacted/crossplane/cmd/crank/version/version.go:51 +0x104
reflect.Value.call({0x105fde320?, 0x1400032f3b8?, 0x14000069aa8?}, {0x105a1bd1e, 0x4}, {0x1400061d9f8, 0x1, 0x104278f88?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/reflect/value.go:581 +0x97c
reflect.Value.Call({0x105fde320?, 0x1400032f3b8?, 0x14000069ba8?}, {0x1400061d9f8?, 0x10637d5c0?, 0x104169d58?})
	/opt/homebrew/Cellar/go/1.23.2/libexec/src/reflect/value.go:365 +0x94
github.com/alecthomas/kong.callFunction({0x105fde320?, 0x1400032f3b8?, 0x14000069d08?}, 0x14000756540)
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/callbacks.go:98 +0x390
github.com/alecthomas/kong.(*Context).RunNode(0x14000817300, 0x140005bb680, {0x0, 0x0, 0x14000574b40?})
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/context.go:774 +0x604
github.com/alecthomas/kong.(*Context).Run(0x105ea8560?, {0x0?, 0x14000069ee8?, 0x10000000000013?})
	/Users/moulick.aggarwal/go/pkg/mod/github.com/alecthomas/kong@v0.9.0/context.go:799 +0x138
main.main()
	/Users/moulick.aggarwal/redacted/crossplane/cmd/crank/main.go:77 +0x240
```

</details>

3. With `-trimpath` additionally

```shell
$ go build -o crank -trimpath -ldflags='-s -w -X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version
```

<details>
<summary>Panic</summary>

```shell
$ go build -o crank -trimpath -ldflags='-s -w -X github.com/crossplane/crossplane/internal/version.version=v1.2.3' ./cmd/crank; ./crank version
Client Version: v1.2.3
panic: Trim the fat test

goroutine 1 [running]:
github.com/crossplane/crossplane/cmd/crank/version.(*Cmd).Run(0x1400004f8b8, 0x14000370680)
	github.com/crossplane/crossplane/cmd/crank/version/version.go:51 +0x104
reflect.Value.call({0x10252a320?, 0x1400004f8b8?, 0x14000069aa8?}, {0x101f67d1e, 0x4}, {0x1400000d428, 0x1, 0x1007c4f88?})
	reflect/value.go:581 +0x97c
reflect.Value.Call({0x10252a320?, 0x1400004f8b8?, 0x14000069ba8?}, {0x1400000d428?, 0x1028c95c0?, 0x1006b5d58?})
	reflect/value.go:365 +0x94
github.com/alecthomas/kong.callFunction({0x10252a320?, 0x1400004f8b8?, 0x14000069d08?}, 0x140006b2120)
	github.com/alecthomas/kong@v0.9.0/callbacks.go:98 +0x390
github.com/alecthomas/kong.(*Context).RunNode(0x14000370680, 0x1400057d0e0, {0x0, 0x0, 0x1400056a3c0?})
	github.com/alecthomas/kong@v0.9.0/context.go:774 +0x604
github.com/alecthomas/kong.(*Context).Run(0x1023f4560?, {0x0?, 0x14000069ee8?, 0x10000000010013?})
	github.com/alecthomas/kong@v0.9.0/context.go:799 +0x138
main.main()
	github.com/crossplane/crossplane/cmd/crank/main.go:77 +0x240
```

</details>

So panic stack trace retains all the information. With the addition of `-trimpath` it also removed my local paths, which on one hand make it slight harder to navigate to the exact file, on the other hand provides privacy and prevent leaking build time paths.

>     * some part of controller-runtime wants to print the stack, like we've seen in multiple places e.g. [v0.44.0 - [controller-runtime] log.SetLogger(...) was never called; logs will not be displayed. crossplane-contrib/provider-upjet-aws#973](https://github.com/crossplane-contrib/provider-upjet-aws/issues/973) - is it still possible to see that stack output?

This one I do not know, but if `panic()` can still print the same trace, it should also be able to do the same?

**Comment by Moulick** [comment]
@jbw976 @negz could you please trigger the GHA so I can fix any issues in parallel?

**Comment by Moulick** [comment]
e2e test failure seems to be a common race condition or parallel test suite type of failure in k8s testing. It'll probably pass on another attempt.

**Comment by Moulick** [comment]
Looking at the stack traces above, I think it should be good to go? Is there something else needed?

**Review by jbw976 — APPROVED** [review]
Awesome, thanks for the thoroughness @Moulick to test how the built binaries will behave with panics. I'm a bit surprised, but definitely pleased, that we still retain the basic stack trace information that a panic outputs. 

I think this is a reasonable change - we decrease the binary size but keep the ability to debug typical failure scenarios. Thanks for figuring this out!

**Pull request merged**